### DISCUSSION TITLE
Color4 uses Vector128?

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#283] [General]
Could you please explain why you use a 128 bit intrinsic type for this 32bit type? My best guess is that you've calculated that even though it's larger, the speed gain from using intrinsics makes it worth while. 

But a consideration for me is that up until recently I've used a Color8 struct (8bpc) - it's very compact. Recently I've started using a Color32 struct (32 bpc) .. because I want to eventually accommodate HDR and  I want to get used to working with colors that way and get everything to support that type. Should I be worried about using these large types (32 bit or even this 128 bit intrinsic type) in vertex geometry and textures. Or not worry? Because it would blow out my data sizes.

I'm also considering whether I should use these intrinsics in my own types. 

What is your thinking (and anyone else who want's to comment is welcome) on using large dataypes for color which are used in large datasets such as geometry and textures.

And why not use Vector4 as the underlying type. Doesn't it also use intrinsics and is more compact?

```
    [StructLayout(LayoutKind.Sequential, Pack = 4)]
    public readonly struct Color4 : IEquatable<Color4>, IFormattable
    {
        private readonly Vector128<float> _value;
```

**Comment by paulbartrum** [comment]
If you're looking for a 32-bit (8-bit per component) color type, it's `Vortice.Mathematics.ColorBgra`. Color4 is explicitly a 128-bit (32-bit per component) color type.

In terms of what is recommended for vertex geometry and textures, you should use 32-bit colors in all cases except where you need high-definition color. An example of where you might need HDR color is for a bright sky texture, since the sun is much brighter than 80 nits. Note that if you're using HDR textures in Direct3D you can reduce the storage and memory requirements by using BC6H compression (and OpenGL/Vulkan has an equivalent option).

**Comment by Gavin-Williams** [comment]
OK, so I may have been confused about the Vector128 type. It's actually the same size as a Vector4.

@paulbartrum thanks for the hdr info. I still have to do a bit of reading on how to handle HDR. It seems like everyone says keep textures in SDR and handle the HDR in shader. Obviously the example you gave is a good one for when to use a HDR texture.

**Reply by paulbartrum** [reply]
Yes it is, and yes that is confusing 🙂

**Reply by amerkoleci** [reply]
Yes, as @paulbartrum described Vector128 float is just _m128 in SSE world or float32x4_t when using ARM intrinsics, so the size is just as Vector4/Color4.
